### PR TITLE
tests, sriov: Wait for VMI removal at test teardown

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

SR-IOV tests execution on CI has shown flakes where the VF configuration
has been modified after assigning the device to the pod.

The primary suspect is a CNI DELETE instruction from a previous pod
which is somehow racing with the re-assignment of the device to the new
pod.

In order to mitigate this suspected issue (and as side effect, prove it)
this change assures the VMI/s used with SR-IOV are deleted and absent
before starting the next test.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
~Depends on #8299 (only the last 2 commits are relevant)~

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: tests/network/sriov.go [hunk 1]
```diff
@@ -697,6 +715,34 @@ func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error)
 	return virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &k8smetav1.GetOptions{})
 }
 
+// deleteVMI deletes the specified VMI and waits for its absence.
+// Waiting for the VMI removal is placed intentionally for VMI/s with SR-IOV networks in order
+// to assure resources (VF/s) are fully released before reused again on a new VMI.
+func deleteVMI(vmi *v1.VirtualMachineInstance) error {
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		panic(err)
```
Comment (oshoval) [inline tests/network/sriov.go:724]: maybe Expect instead?

File: tests/network/sriov.go [hunk 2]
```diff
@@ -697,6 +715,34 @@ func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error)
...
+
+	const timeout = 30 * time.Second
+	err = wait.Poll(1*time.Second, timeout, func() (done bool, err error) {
+		_, getErr := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &k8smetav1.GetOptions{})
+		if k8serrors.IsNotFound(getErr) {
+			return true, nil
+		}
+		return false, getErr
+	})
+	return fmt.Errorf("failed to wait for the VMI to disapear: %v", err)
```
Comment (oshoval) [inline tests/network/sriov.go:743]: need if `err != nil` I think ?

**Review by oshoval — COMMENTED** [review]
Thanks
(reviewed only last 2 commits)

File: tests/network/sriov.go (hunk 2 above)
Comment (maiqueb) [inline tests/network/sriov.go:743]: nit: I'd find this easier to read if it were simply:
```golang
return wait.Poll(1*time.Second, timeout, func() (done bool, err error) {
		_, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &k8smetav1.GetOptions{})
		if k8serrors.IsNotFound(err) {
			return true, nil
		}
		return false, err
	})
```

**Comment by oshoval** [comment]
Can we use `AfterEach` to remove the VMI ?
If the name that is used is predictable it might work (as the function robust to "not found" error upon deletion)

**Comment by xpivarc** [comment]
@EdDev Why does this depends on https://github.com/kubevirt/kubevirt/pull/8299 ? This would greatly benefit the CI right now as far as I understand.

**Comment by EdDev** [comment]
> @EdDev Why does this depends on #8299 ? This would greatly benefit the CI right now as far as I understand.

I found it harder to do this on a messy code base.
I did not expect the refactoring to be so dramatic to be an issue.

File: tests/network/sriov.go (hunk 1 above)
Comment (EdDev) [reply tests/network/sriov.go:724]: Client fetching has always been treated with panic.
This has been also raised in the refactoring PR, and the answer is the same.. Failing to fetch the client is a test failure, not the tested subject error.

File: tests/network/sriov.go (hunk 2 above)
Comment (EdDev) [reply tests/network/sriov.go:743]: Done

File: tests/network/sriov.go (hunk 1 above)
Comment (oshoval) [reply tests/network/sriov.go:724]: Note: I see mixed styles fwiw
sometimes panic, sometimes Expect
and even one time it just skip tests (wrong) here `GetKubernetesVersion`

File: tests/network/sriov.go (hunk 2 above)
Comment (EdDev) [reply tests/network/sriov.go:743]: I wanted to differentiate between the waiting timeout error and the error of the deleting action.
Do you think it is not needed?

**Comment by EdDev** [comment]
change: Fixed a bug based on a [comment](https://github.com/kubevirt/kubevirt/pull/8307/commits/2425660d61f1e1ad75e021076da5dc92751fc90c#r948971162).

**Comment by EdDev** [comment]
> Can we use `AfterEach` to remove the VMI ? (instead defer on each test) If the name that is used is predictable it might work (as the function robust to "not found" error upon deletion)

It will be ugly IMO. Using defer only when needed and close to the creation makes more sense to me.

File: tests/network/sriov.go (hunk 2 above)
Comment (maiqueb) [reply tests/network/sriov.go]: Doesn't `wait.Poll` [return an error](https://github.com/kubernetes/apimachinery/blob/97e5df2d0258ac077093867fabf508c1fc31f53b/pkg/util/wait/wait.go#L592) if it times out ? 

If so, the thrown error will just be different (timeout vs error in the caller), won't it ?

**Comment by maiqueb** [comment]
> [change](https://github.com/kubevirt/kubevirt/compare/a72120d815567109cb98a6b0e4420f0f4ee1694e..886a428e31e6704bef3c79f7d3b8f0ad110d04ed): Fixed a bug based on a [comment](https://github.com/kubevirt/kubevirt/pull/8307/commits/2425660d61f1e1ad75e021076da5dc92751fc90c#r948971162).

Unfortunately, the link to the `comment` now points to nowhere - forced push removed that commit. Could you link to the "comment" instead ? @EdDev 

Thanks in advance .

**Comment by xpivarc** [comment]
> > @EdDev Why does this depends on #8299 ? This would greatly benefit the CI right now as far as I understand.
> 
> I found it harder to do this on a messy code base. I did not expect the refactoring to be so dramatic to be an issue.

So I don't mind refactoring the code with actual changes as long as it helps. Here I would like to remove the refactor changes as these changes are obliviously taking more time to review. I also dislike we are running CI on 2 PRs which are less or more the same... and the reviews are splitting into 2 places.

**Comment by EdDev** [comment]
> > [change](https://github.com/kubevirt/kubevirt/compare/a72120d815567109cb98a6b0e4420f0f4ee1694e..886a428e31e6704bef3c79f7d3b8f0ad110d04ed): Fixed a bug based on a [comment](https://github.com/kubevirt/kubevirt/pull/8307/commits/2425660d61f1e1ad75e021076da5dc92751fc90c#r948971162).
> 
> Unfortunately, the link to the `comment` now points to nowhere - forced push removed that commit. Could you link to the "comment" instead ? @EdDev

I tried to link to the comment.
Lets try again: https://github.com/kubevirt/kubevirt/pull/8307#discussion_r948971162

File: tests/network/sriov.go (hunk 2 above)
Comment (EdDev) [reply tests/network/sriov.go]: I think you are right then... lets do this.

**Comment by EdDev** [comment]
> So I don't mind refactoring the code with actual changes as long as it helps. Here I would like to remove the refactor changes as these changes are obliviously taking more time to review.

Yes, it requires more review time. But if I had to write the same in the old code, maybe the reviewers will need more effort to assure it is correct.

For me, tt is just a personal ability to work in a messy code vs in a cleaner one. I find it easier to fix the mess and then do the logical change. It makes me feel safer and from my experience, it reduces the chances of introducing bugs.
For me, it is more work to work in the original code and directly do this fix.

> I also dislike we are running CI on 2 PRs which are less or more the same... and the reviews are splitting into 2 places.

The two PR/s are doing two different things, and the review is indeed focusing on two different things. One about refactoring and the other about adding the cleanup logic.
Discussing both topics in one PR will overload the discussion and make the review harder IMO.

File: tests/network/sriov.go (hunk 2 above)
Comment (oshoval) [reply tests/network/sriov.go]: Please consider using `PollImmediate`, it doesn't wait the interval before checking the condition in the first time from what i understand.

Comment (EdDev) [reply tests/network/sriov.go]: Done

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/886a428e31e6704bef3c79f7d3b8f0ad110d04ed..2d0cec3f1dd2ea7cfac402f87061a51d6d0c84b7): Answered the [comment](https://github.com/kubevirt/kubevirt/pull/8307#discussion_r949125630) from the review.

File: tests/network/sriov.go [hunk 3]
```diff
@@ -102,6 +103,7 @@ var _ = Describe("[Serial]SRIOV", func() {
 			var err error
 			vmi, err = createAndWaitVMI(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			DeferCleanup(deleteVMI, vmi)
```
Comment (oshoval) [inline tests/network/sriov.go:106]: Lets say the VM failed to be deleted, will we see the error when there is only return there?
If not, would Expect would help ?

File: tests/network/sriov.go (hunk 2 above)
Comment (EdDev) [reply tests/network/sriov.go]: > While changing, please consider using `PollImmediate`, it doesn't wait the interval before checking the condition in the first time from what i understand.

Fine by me. Done

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/2d0cec3f1dd2ea7cfac402f87061a51d6d0c84b7..cf4bab74b8e9ad3ddcb4ec399325579ce5329533): Answered comment regarding `Poll`.

File: tests/network/sriov.go (hunk 3 above)
Comment (EdDev) [reply tests/network/sriov.go:106]: It creates a dynamic `AfterEach`, so the answer is yes.

**Comment by xpivarc** [comment]
>Depends on https://github.com/kubevirt/kubevirt/pull/8299 (only the last 2 commits are relevant)
/hold

File: tests/network/sriov.go (hunk 3 above)
Comment (oshoval) [reply tests/network/sriov.go:106]: but there is just a return in that dynamic `AfterEach`, don't we need `Expect` so it will be visible ?
(talking about the last return, not the panic)

File: tests/network/sriov.go (hunk 1 above)
Comment (maiqueb) [reply tests/network/sriov.go:724]: I think panicking might be a better action to take, since I think there's little chance the subsequent tests can create a working client afterwards.

File: tests/network/sriov.go [hunk 4]
```diff
@@ -695,3 +657,157 @@ func findIfaceByMAC(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineIns
...
+	if err != nil {
+		panic(err)
+	}
+
+	tests.WaitAgentConnected(virtClient, vmi)
+
+	return virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &k8smetav1.GetOptions{})
+}
+
+// deleteVMI deletes the specified VMI and waits for its absence.
```
Comment (xpivarc) [inline tests/network/sriov.go:718]: My understanding is this is a bug so please let's ensure we have an issue, todo... to remove this hack in the future.

File: tests/network/sriov.go (hunk 3 above)
Comment (oshoval) [inline tests/network/sriov.go:106]: Why do we need `DeferCleanup` ? why not just defer?
The first is context aware that is needed for example on `BeforeEach` (where it will auto create an `AfterEach` as you said)

File: tests/network/sriov.go (hunk 4 above)
Comment (EdDev) [reply tests/network/sriov.go:718]: I consider the global cleanup to be wrong, so I consider this to be the right way to write tests.
But if you insist, I will add it.

Comment (oshoval) [reply tests/network/sriov.go:718]: I am with Edy here, we never block upon delete, its not PR specific

**Comment by EdDev** [comment]
@oshoval , please try to do the review once and post all you comments as a group.
Receiving individual comments every few minutes is making it harder for me to answer.

File: tests/network/sriov.go (hunk 3 above)
Comment (EdDev) [reply tests/network/sriov.go:106]: A simple `defer` will not allow the reporter to collect the state of the system once an error occurs in the test.

**Comment by oshoval** [comment]
> @oshoval , please try to do the review once and post all you comments as a group. Receiving individual comments every few minutes is making it harder for me to answer.

Usually i do them in a group, but honestly i thought about them just afterwards
Will try to buffer

File: tests/network/sriov.go (hunk 4 above)
Comment (xpivarc) [reply tests/network/sriov.go:718]: That is another topic and we should solve it in a general way imho. WDYT?

**Comment by oshoval** [comment]
Conceptual lgtm for the 2 last commits,
Thanks

File: tests/network/sriov.go (hunk 4 above)
Comment (EdDev) [reply tests/network/sriov.go:718]: The question remains:
> Should we wait for all VMI/s to be gone before declaring the test as finished?

I argue we should, in all cases, making sure we do not start a test while older objects exist in the cluster, which may have unclear expectations.

But for the SR-IOV tests specifically, I think the requirement is even stronger.
The resources each VMI uses are HW related and keeping old ones around may causes issues, like the one we see now.
It is true that it may occur due to a bug, but we should asks ourself if this stress test was indeed explicitly intended or just a side effect. I think we can agree it is a side-effect.

The only general rule I can think of, is this: If a test creates objects that use cluster or node wide resources, it will be correct to wait for the object removal before declaring the test over.
As SR-IOV VMI/s answer this at the node resource level, I think the suggested change here can be consider a good solution, both to the current problem we sometime face and to the general proposed "rule".

But again, if you are not convinced, and this is will be reason to not push this PR forward, I will add the ref to a the bug.

**Comment by EdDev** [comment]
I think at this stage the CI is not blocked by what this PR tries to fix. The occurrences of the error on the SR-IOV jobs have not been seen in the last week.
If they re-appear at high rate, we need to re-evaluate the urgency here.

File: tests/network/sriov.go (hunk 4 above)
Comment (maiqueb) [reply tests/network/sriov.go:718]: All of these are imo requirements for the testing framework @fossedihelm proposed - some stuff can be lazily deleted (assuming we'd be using a different namespace for each test) but some - like sriov resources - **must** be synchronously deleted on the test teardown.

Are these scenarios considered on your proposal @fossedihelm ? 

... I'm aware I owe you a review on that design doc ...

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/cf4bab74b8e9ad3ddcb4ec399325579ce5329533..c480605578bfead9c38e6959db1deb877646dde7): Rebase

**Comment by EdDev** [comment]
> > Depends on #8299 (only the last 2 commits are relevant)
> 
> /hold

/unhold

**Review by oshoval — APPROVED** [review]
Thanks

Please remove
"Depends on https://github.com/kubevirt/kubevirt/pull/8299 (only the last 2 commits are relevant)"

File: tests/network/sriov.go [hunk 5]
```diff
@@ -689,6 +707,33 @@ func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error)
...
+// Waiting for the VMI removal is placed intentionally for VMI/s with SR-IOV networks in order
+// to assure resources (VF/s) are fully released before reused again on a new VMI.
+func deleteVMI(vmi *v1.VirtualMachineInstance) error {
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		panic(err)
+	}
+
+	GinkgoWriter.Println("sriov:", vmi.Name, "deletion started")
```
Comment (orelmisan) [inline tests/network/sriov.go:719]: @EdDev WDYT about adding `vmi.Namespace` to this log entry as well?

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @EdDev.

**Comment by AlonaKaplan** [comment]
> **What this PR does / why we need it**:
> 
> SR-IOV tests execution on CI has shown flakes where the VF configuration has been modified after assigning the device to the pod.
> 
> The primary suspect is a CNI DELETE instruction from a previous pod which is somehow racing with the re-assignment of the device to the new pod.
> 
> In order to mitigate this suspected issue (and as side effect, prove it) this change assures the VMI/s used with SR-IOV are deleted and absent before starting the next test.
> 
> **Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_: Fixes #
> 
> **Special notes for your reviewer**: ~Depends on #8299 (only the last 2 commits are relevant)~
> 
> **Release note**:
> 
> ```
> NONE
> ```

If it is the case, the code itself (not tests) have to be changed to avoid assigning a VF that was previously used by a pod that is still not completely removed. Is there a follow-up task for it?

File: tests/network/sriov.go (hunk 4 above)
Comment (fossedihelm) [reply tests/network/sriov.go:710]: @maiqueb I didn't consider these scenarios, so we should expose a way to set a specific behavior: Wait or don't wait resources deletion. Thanks for pointing me this!

**Comment by EdDev** [comment]
> If it is the case, the code itself (not tests) have to be changed to avoid assigning a VF that was previously used by a pod that is still not completely removed. Is there a follow-up task for it?

@ormergi is working to recreate it on pure k8s and open an issue against the SR-IOV components (operator/dp/cni).
There is nothing we can do on our side (kubevirt).

File: tests/network/sriov.go (hunk 5 above)
Comment (EdDev) [reply tests/network/sriov.go:719]: The name is unique and there is only one ns for the sriov tests. I think it is refundant.

Comment (orelmisan) [reply tests/network/sriov.go:719]: Thanks.

**Review by orelmisan — APPROVED** [review]

File: tests/network/sriov.go (hunk 4 above)
Comment (AlonaKaplan) [reply tests/network/sriov.go:710]: > My understanding is this is a bug so please let's ensure we have an issue, todo... to remove this hack in the future.

I agree with this comment. There is a real bug and we are covering it with this cleanup. Please add a comment with a reference to the bug opened on the relevant component (SRIOV operator or whatever component). If it is still under investigation what is the corresponding component, please open a kubevirt issue.
Even though it is a corner case, it is still a real bug.

File: tests/network/sriov.go [hunk 6]
```diff
@@ -689,6 +707,33 @@ func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error)
...
+	if err = virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &k8smetav1.DeleteOptions{}); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to delete VMI: %v", err)
+	}
+
+	const timeout = 30 * time.Second
+	return wait.PollImmediate(1*time.Second, timeout, func() (done bool, err error) {
```
Comment (AlonaKaplan) [inline tests/network/sriov.go:728]: Please reuse `utils.WaitForVirtualMachineToDisappearWithTimeout`.

File: tests/utils.go [hunk 7]
```diff
@@ -128,11 +128,13 @@ const MigrationWaitTime = 240
 const ContainerCompletionWaitTime = 60
 
 func TestCleanup() {
+	GinkgoWriter.Println("Global test cleanup started.")
```
Comment (AlonaKaplan) [inline tests/utils.go:131]: Usually we use `By` for this purpose. What is the advantage of `GinkgoWriter.Println` over `By`?

File: tests/network/sriov.go (hunk 4 above)
Comment (EdDev) [reply tests/network/sriov.go:710]: We suspect there is a bug caused by this behavior. But it does not matter, our tests are not here to test this scenario (load of reallocations) and cleaning up properly as suggested here is the correct thing, even without this potential bug.

adding such a ref implies we can leave vmi/s around in “deleting” state, which is IMO wrong.

File: tests/utils.go (hunk 7 above)
Comment (EdDev) [reply tests/utils.go:131]: The documentation states that `By` purpose is to split a test into blocks.
I thought it is less of a fit in this case. My interpretation of tge docs is that it fits in the tests (It).

File: tests/network/sriov.go (hunk 4 above)
Comment (AlonaKaplan) [reply tests/network/sriov.go:710]: Yes, our tests are designed this way. A VM from a previous test can be in deleting state when we run a new test.
In most of the tests we are not waiting for the previous VMs to be deleted.
Here we need to wait due to a bug and not because it is the correct test design.
Therefore, we should document it is a workaround.
And if you don't consider it as a workaround then the official documentation should be updated to tell the user has to wait for a previous VM using a resouce to be removed before running a new VM using/potentially using the same resource.

File: tests/network/sriov.go (hunk 6 above)
Comment (AlonaKaplan) [reply tests/network/sriov.go:728]: If this is the case, you can move this helper to a more proper package and add offset parameter to it.
It doesn't justify a copy paste.

File: tests/network/sriov.go (hunk 4 above)
Comment (xpivarc) [reply tests/network/sriov.go:710]: Let me just add a few points:
1. This is a difference in how we do clean-up in the test suite. It is irelevant if this is the "right" way to do it or not.
2. We would not be making this change if we would not hit this bug. Hiding this fact doesn't do right to anyone.

File: tests/network/sriov.go (hunk 6 above)
Comment (EdDev) [reply tests/network/sriov.go:728]: It is a different approach, not a copy/paste.
One asserts in helpers, the other is pure functional and returns errors.
One depends on ginkgo & gomega, the other does not.
One is dependent on offsets, the other does not.

Are you asking to refactor `WaitForVirtualMachineToDisappearWithTimeout` or extracting with "wait" implementation to a function and placing it in a new package?

Comment (AlonaKaplan) [reply tests/network/sriov.go:728]: > Are you asking to refactor WaitForVirtualMachineToDisappearWithTimeout or extracting the "wait.Poll*" implementation to a function and placing it in a new package?

I"m o with both of the options.

File: tests/network/sriov.go (hunk 4 above)
Comment (xpivarc) [reply tests/network/sriov.go:710]: I would personally be for having it here. @AlonaKaplan I let you decide.

Comment (ormergi) [reply tests/network/sriov.go:710]: Tracking issue https://github.com/k8snetworkplumbingwg/sriov-cni/issues/219.

As a side note, I agree that the way VMIs are being cleaned up should be revisited.
The fact that we dispose VMIs by removing their finalizers is in general bad practice and could lead to resource leaks.
https://github.com/kubevirt/kubevirt/blob/28630a3f91434203f2eac3f4c5bce217e6a47833/tests/testsuite/namespace.go#L90-L101

Comment (AlonaKaplan) [reply tests/network/sriov.go:710]: I believe that the correctness of our current cleanup code is irrelevant here.
In this case it helped us find a real bug, so maybe it is not that bad?

The tests here were changed to have a non-standard cleanup due to a bug.
Please add the reference to the bug to the tests code (and not just the commit message) so it will be easier for the next developer that touches it to decide whether the workaround is still needed or not.

Comment (EdDev) [reply tests/network/sriov.go:710]: I added the reference.
Hope this is enough to close this point.

File: tests/network/sriov.go (hunk 6 above)
Comment (EdDev) [reply tests/network/sriov.go:729]: I opened a follow up PR for this: https://github.com/kubevirt/kubevirt/pull/8365
